### ISSUE TITLE
Incompatible output.xml created if listener runs keyword in `end_keyword` inside FOR loop

### ISSUE THREAD

**Issue by Harri** [robotframework/robotframework#4112]
Given following test suite:

    *** Test Cases ***
    List
        ${list}    Create List    foo    bar
        FOR    ${element}    IN    @{list}
            Log    ${element}
            Fail    Error
        END

    *** Keywords ***
    Log Error Situation
        Log    Error Situation

and following `listener.py`

    from robot.libraries.BuiltIn import BuiltIn

    ROBOT_LISTENER_API_VERSION = 2

    def end_keyword(name, attributes):
        if attributes['status'] == 'FAIL':
            BuiltIn().run_keyword('Log Error Situation')

When executing `robot --listener .\listener.py .` We get this output

    > robot --listener .\listener.py .
    ==============================================================================
    Rf-For
    ==============================================================================
    Rf-For.Test
    ==============================================================================
    List                                                                  | FAIL |
    Error
    ------------------------------------------------------------------------------
    Rf-For.Test                                                           | FAIL |
    1 test, 0 passed, 1 failed
    ==============================================================================
    Rf-For                                                                | FAIL |
    1 test, 0 passed, 1 failed
    ==============================================================================
    Output:  C:\Users\hpaavola\Development\sandbox\rf-for\output.xml
    [ ERROR ] Reading XML source 'C:\Users\hpaavola\Development\sandbox\rf-for\output.xml' failed: Incompatible child element 'kw' for 'for'.

Try --help for usage information.

**Comment by pekkaklarck** [comment]
The problem is that listener is calling a keyword in a place where keywords typically cannot exist. It creates a `<kw>` element in output.xml where it's not expected and our output.xml parsing code fails. PR #4111 fixes this by disabling listeners in this particular case. I don't think that's a good idea because there can be legitimate reasons to get a listener called. And listeners typically don't run keywords and thus cause problems like this.

It's probably a better idea to fix output.xml processing to handle `<kw>` inside `<for>`. Luckily this code is pretty generic and fixing the problem requires only this:

```diff
diff --git a/src/robot/result/xmlelementhandlers.py b/src/robot/result/xmlelementhandlers.py
index ae4b1d075..f8e4b046d 100644
--- a/src/robot/result/xmlelementhandlers.py
+++ b/src/robot/result/xmlelementhandlers.py
@@ -168,7 +168,7 @@ class KeywordHandler(ElementHandler):
 @ElementHandler.register
 class ForHandler(ElementHandler):
     tag = 'for'
-    children = frozenset(('var', 'value', 'doc', 'status', 'iter', 'msg'))
+    children = frozenset(('var', 'value', 'doc', 'status', 'iter', 'msg', 'kw'))
 
     def start(self, elem, result):
         return result.body.create_for(flavor=elem.get('flavor'))
```

The above isn't enough yet, because the `ForIterations` model object doesn't support adding keywords into its body, it only can contain `ForIteration` and `Message` objects. Luckily also this code is pretty generic and fixing this problem just requires removing one line of code:

```diff
diff --git a/src/robot/result/model.py b/src/robot/result/model.py
index d00dd2681..069ee845b 100644
--- a/src/robot/result/model.py
+++ b/src/robot/result/model.py
@@ -63,7 +63,6 @@ class Body(model.Body):
 
 class ForIterations(Body):
     for_iteration_class = None
-    keyword_class = None
     if_class = None
     for_class = None
     __slots__ = []
```

The above fix causes one unit test to fail, a test explicitly testing `ForIterations.create_keyword` doesn't work, so that test needs to be fixed. Probably should be changed to test that `ForIterations.create_keyword` works. This whole fix will then also needs acceptance tests.

Are you @Harri interested to take care of the above? I can also do that tomorrow and then create RF 4.1.2 rc 2 with the fix.

**Comment by pekkaklarck** [comment]
Although this is severe issue, I decided to give it medium priority because it occurs only in pretty special usage and because there's a pretty easy way to avoid it. Basically the listener can check the type of the started "keyword" and avoid FOR loops. Severity doesn't really matter, though, as I anyway want this to be fixed in RF 4.1.2.

**Comment by pekkaklarck** [comment]
One more thing: Need to check that our output.xml schema considers `<kw>` inside `<for>` valid. Our acceptance test system ought to validate that automatically, but it's still a good idea to take a look at the schema itself.

**Comment by pekkaklarck** [comment]
That error is due to our acceptance test system adding extra attributes to our model objects. It fails because the original `Keyword` class uses `__slots__` which makes it impossible to add such extra attributes. Our atest system works around that by replacing original classes with "no-slots" variants but that code apparently doesn't take it into account that `ForIterations` now also has keywords. This change ought to fix that:
```diff
diff --git a/atest/resources/TestCheckerLibrary.py b/atest/resources/TestCheckerLibrary.py
index 239f6c492..b55018eea 100644
--- a/atest/resources/TestCheckerLibrary.py
+++ b/atest/resources/TestCheckerLibrary.py
@@ -44,6 +44,7 @@ class NoSlotsForIteration(ForIteration):
 
 class NoSlotsForIterations(ForIterations):
     for_iteration_class = NoSlotsForIteration
+    keyword_class = NoSlotsKeyword
 
 
 NoSlotsKeyword.body_class = NoSlotsBody
```

**Comment by pekkaklarck** [comment]
Are you @Harri working with this? I got some nice performance enhancements done (see #4114) that require a new RF 4.1.2 release candidate. I'd like to create that latest on Friday afternoon and this fix need to be included to make it into RF 4.2.2 final. If you can create PR that's awesome, if not, I can take care of the fix myself.

**Comment by pekkaklarck** [comment]
No worries. Fixed now and I'll create rc2 later today. Please test it and report possible problems!
